### PR TITLE
Hanging Entities can now inherit reinforcements

### PR DESCRIPTION
**Pull request by Protonull**
Hanging entities such as Item Frames can now be protected by their host block's reinforcement

### CODE REVIEW

File: src/vg/civcraft/mc/citadel/listener/EntityListener.java [hunk 1]
```diff
@@ -149,9 +150,43 @@ public void playerQuitEvent(PlayerQuitEvent event){
...
+			}
+			PermissionType permission = PermissionType.getPermission("REINFORCE");
+			// If the REINFORCE permission is not registered, do nothing
+			if (permission == null) {
+				Citadel.getInstance().warning("Could not get the REINFORCE permission from NameLayer. Is it loaded?");
+				return;
+			}
+			Player player = event.getPlayer();
+			// If the player is a member of the group and has bypass permissions, do nothing
+			if (group.isMember(player.getUniqueId()) &&  gm.hasAccess(group, player.getUniqueId(), permission)) {
```
Comment (Maxopoly) [inline src/vg/civcraft/mc/citadel/listener/EntityListener.java]: Don't need the member check here, only the perms one is good enough

File: src/vg/civcraft/mc/citadel/listener/EntityListener.java [hunk 2]
```diff
@@ -149,9 +150,43 @@ public void playerQuitEvent(PlayerQuitEvent event){
...
+				return;
+			}
+			// Otherwise prevent the player from putting item frames on other people's reinforced blocks
+			player.sendMessage(ChatColor.RED + "You cannot place those on blocks you don't have permissions for.");
+			event.setCancelled(true);
+			Bukkit.getScheduler().runTaskLater(Citadel.getInstance(), player::updateInventory, 1L);
```
Comment (Maxopoly) [inline src/vg/civcraft/mc/citadel/listener/EntityListener.java]: Fancy method references

File: src/vg/civcraft/mc/citadel/listener/EntityListener.java [hunk 3]
```diff
@@ -310,10 +396,43 @@ public void hangingEntityBreakEvent(HangingBreakByEntityEvent event){
...
+			}
+			PlayerReinforcement playerReinforcement = (PlayerReinforcement) reinforcement;
+			Group group = playerReinforcement.getGroup();
+			// If the player reinforcement doesn't have a group, do nothing
+			if (group == null) {
+				return;
+			}
+			Player player = event.getPlayer();
+			// If the player is a member of the group and has bypass permissions, do nothing
+			if (group.isMember(player.getUniqueId()) && playerReinforcement.canBypass(player)) {
```
Comment (Maxopoly) [inline src/vg/civcraft/mc/citadel/listener/EntityListener.java]: Dont need a member check here either

File: src/vg/civcraft/mc/citadel/listener/EntityListener.java [hunk 4]
```diff
@@ -310,10 +396,43 @@ public void hangingEntityBreakEvent(HangingBreakByEntityEvent event){
...
+				return;
+			}
+			// Otherwise prevent interaction and notify the player they do not have perms
+			player.sendRawMessage(ChatColor.RED + "You do not have permission to alter that.");
```
Comment (Maxopoly) [inline src/vg/civcraft/mc/citadel/listener/EntityListener.java]: Should probably just be sendMessage

File: src/vg/civcraft/mc/citadel/listener/EntityListener.java [hunk 5]
```diff
@@ -327,9 +446,48 @@ public void playerEntityInteractEvent(PlayerInteractEntityEvent event){
...
+			}
+			PlayerReinforcement playerReinforcement = (PlayerReinforcement) reinforcement;
+			Group group = playerReinforcement.getGroup();
+			// If the player reinforcement somehow does not have a group, just back out
+			if (group == null) {
+				return;
+			}
+			Player player = (Player) event.getDamager();
+			// If the player is a member of the group and has bypass permissions, do nothing
+			if (group.isMember(player.getUniqueId()) && playerReinforcement.canBypass(player)) {
```
Comment (Maxopoly) [inline src/vg/civcraft/mc/citadel/listener/EntityListener.java]: No member check needed here either

File: src/vg/civcraft/mc/citadel/listener/EntityListener.java [hunk 6]
```diff
@@ -327,9 +446,48 @@ public void playerEntityInteractEvent(PlayerInteractEntityEvent event){
...
+				return;
+			}
+			// Otherwise prevent interaction and notify the player they do not have perms
+			player.sendRawMessage(ChatColor.RED + "The host block is protecting this.");
```
Comment (Maxopoly) [inline src/vg/civcraft/mc/citadel/listener/EntityListener.java]: sendMessage() here as well

**Review by Maxopoly — CHANGES_REQUESTED** [review]
We likely want to remove the ability to reinforce ItemFrames entirely eventually, but that doesnt have to be here

File: src/vg/civcraft/mc/citadel/listener/EntityListener.java (hunk 4 above)
Comment (Protonull) [reply src/vg/civcraft/mc/citadel/listener/EntityListener.java]: Oops, damn you Intellij

File: src/vg/civcraft/mc/citadel/listener/EntityListener.java (hunk 2 above)
Comment (Protonull) [reply src/vg/civcraft/mc/citadel/listener/EntityListener.java]: Thanks :)

**Comment by Protonull** [comment]
Done, I also removed the Item Frame reinforcements code entirely as you suggested

**Pull request merged**